napari/utils/progress.py

```diff
@@ -0,0 +1,315 @@
+
+    See tqdm.tqdm API for valid args and kwargs: https://tqdm.github.io/docs/tqdm/
+
+    Also, any keyword arguments to the :class:`ProgressBar` `QWidget`
+    are also accepted and will be passed to the ``ProgressBar``.
+
+    Examples
+    --------
+
+    >>> def long_running(steps=10, delay=0.1):
+    ...     for i in progress(range(steps)):
+    ...         sleep(delay)
+
+    it can also be used as a context manager:
+
+    >>> def long_running(steps=10, repeats=4, delay=0.1):
+    ...     with progress(range(steps)) as pbr:
+    ...         for i in pbr:
+    ...             sleep(delay)
+
+    or for manual updates:
+
+    >>> def manual_updates(total):
+    ...     pbr = progress(total=total)
+    ...     sleep(10)
+    ...     pbr.set_description("Step 1 Complete")
+    ...     pbr.update(1)
+
+    """
+
+    def __init__(
+        self,
+        iterable: Optional[Iterable] = None,
+        desc: Optional[str] = None,
+        total: Optional[int] = None,
+        *args,
+        **kwargs,
+    ) -> None:
+
+        # check if there's a napari viewer instance
+        viewer = get_viewer_instance()
+        if viewer is not None:
+            kwargs['gui'] = True
+
+        kwargs = kwargs.copy()
+        pbar_kwargs = {k: kwargs.pop(k) for k in set(kwargs) - _tqdm_kwargs}
+
+        super().__init__(iterable, desc, total, *args, **kwargs)
+        if viewer is None:
+            return
+        self.viewer = viewer
+
+        self._pbar = get_pbar(viewer, **pbar_kwargs)
+        if self.total is not None:
+            self._pbar.setRange(self.n, self.total)
+            self._pbar._set_value(self.n)
+        else:
+            self._pbar.setRange(0, 0)
+
+        if desc:
+            self.set_description(desc)
+        else:
+            desc = get_calling_function_name(max_depth=5)
+            if desc:
+                self.set_description(desc)
+            else:
+                # TODO: pick a better default
+                self.set_description("Progress Bar")
+
+        self.show()
+        QApplication.processEvents()
+
+    def display(self, msg: str = None, pos: int = None) -> None:
+        """Update the display."""
+        if not self.viewer:
+            return super().display(msg=msg, pos=pos)
+
+        eta_params = {
+            k: self.format_dict[k]
+            for k in [
+                'n',
+                'total',
+                'elapsed',
+                'unit',
+                'unit_scale',
+                'rate',
+                'unit_divisor',
+                'initial',
+            ]
+        }
+        etas = self.format_time(**eta_params)
+
+        self._pbar._set_value(self.n)
+        self._pbar._set_eta(etas)
+        QApplication.processEvents()
+
+    def set_description(self, desc):
+        """Update progress bar description"""
+        super().set_description(desc, refresh=True)
+
+        self._pbar._set_description(self.desc)
+
+    @staticmethod
+    def format_time(
```

Comment:
this feels like a lot of code... does it do more than just strip the "bar" part from the `format_meter`?  could you perhaps just take the output of `format_meter` and split it or otherwise trim it to your liking?  or are there possibly some variables that modify the output of `format_meter`?

Reply:
After looking closer I think we can just do this:

```
str(self).split('|')[-1]
```

This calls `format_meter` for us and gives us just the ETA part. Even if someone passes a custom `bar_format`, `format_meter` still adds the `"|"`, and if somehow someone is using this without a `total` or an `iterable` we just get the full string we already were from `format_meter`.

We might want to provide a way for people to set their own ETAs? But maybe that's overkill..